controllers/infrastructure/k8sinstallerconfig_controller.go

```diff
@@ -0,0 +1,317 @@
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.2/pkg/reconcile
+func (r *K8sInstallerConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconcile request received")
+
+	// Fetch the K8sInstallerConfig instance
+	config := &infrav1.K8sInstallerConfig{}
+	err := r.Client.Get(ctx, req.NamespacedName, config)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "Failed to get K8sInstallerConfig")
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the ByoMachine
+	byoMachine, err := GetOwnerByoMachine(ctx, r.Client, &config.ObjectMeta)
+	if err != nil {
+		logger.Error(err, "failed to get Owner ByoMachine")
+		return ctrl.Result{}, err
+	}
+
+	if byoMachine == nil {
+		logger.Info("Waiting for ByoMachine Controller to set OwnerRef on InstallerConfig")
+		return ctrl.Result{}, nil
+	}
+	logger.Info("byoMachine found")
+
+	// Fetch the Cluster
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, byoMachine.ObjectMeta)
+	if err != nil {
+		logger.Error(err, "ByoMachine owner Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, err
+	}
+
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	if annotations.IsPaused(cluster, config) {
+		logger.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Create the K8sInstallerConfig scope
+	scope := &k8sInstallerConfigScope{
+		Client:     r.Client,
+		Logger:     logger,
+		Cluster:    cluster,
+		ByoMachine: byoMachine,
+		Config:     config,
+	}
+
+	helper, err := patch.NewHelper(config, r.Client)
+	if err != nil {
+		logger.Error(err, "unable to create helper")
+		return ctrl.Result{}, err
+	}
+	defer func() {
+		if err = helper.Patch(ctx, config); err != nil && reterr == nil {
+			logger.Error(err, "failed to patch K8sInstallerConfig")
+			reterr = err
+		}
+	}()
+
+	// Add finalizer first if not exist
+	if !controllerutil.ContainsFinalizer(scope.Config, infrav1.K8sInstallerConfigFinalizer) {
+		controllerutil.AddFinalizer(scope.Config, infrav1.K8sInstallerConfigFinalizer)
+	}
+
+	// Handle deleted K8sInstallerConfig
+	if !config.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, scope)
+	}
+
+	switch {
+	// waiting for ByoMachine to updating it's ByoHostReady condition to false for reason InstallationSecretNotAvailableReason
+	case conditions.GetReason(byoMachine, infrav1.BYOHostReady) != infrav1.InstallationSecretNotAvailableReason:
+		logger.Info("ByoMachine is not waiting for InstallationSecret", "reason", conditions.GetReason(byoMachine, infrav1.BYOHostReady))
+		return ctrl.Result{}, nil
+	// Status is ready means a config has been generated.
+	case config.Status.Ready:
+		logger.Info("K8sInstallerConfig is ready")
+		return ctrl.Result{}, nil
+	}
```

Comment:
nitpick: Do you think that `if` statements could be more redable here.

Reply:
I think, readability if both are same only.